### ISSUE TITLE
Possible ways to speedup e2e testing?

### ISSUE THREAD

**Issue by wusatosi** [Submitty/Submitty#6390]
Currently, e2e testing runs for about an hour on each commit. This will worsen as we introduce more and more features to Submitty. I've had a look around to see all the logs. This is what I've found:

Top three actions ordered by time spent, cumulative to ~50 mins of each run.
| Order | Name                                 | Time spent | Related Scripts                    |
|-------|--------------------------------------|------------|------------------------------------|
| 1     | Run integration tests                | ~30min     | test cases under `tests/integrationTests/tests/`  driver:  `tests/integrationTests/lib.py`   |
| 2     | Set up sample course                 | ~15min     | `.setup/bin/setup_sample_courses.py` |
| 3     | Run selenium e2e tests with pam auth | ~10min     | test cases under `tests/e2e/*.py`    (using built in unittest driver)      |

# Proposing solution: (high-level overview) 
## Parallel testing
All the tests currently run sequentially, observing from the logs, there is no parallel testing implemented yet. Since testing are (suppose to be) completely isolated, this should be fairly easy to parallelize. 

### Integration testing:
For integration testing, we have a custom wrote driver. The main loop is implemented here:
https://github.com/Submitty/Submitty/blob/dbee91d58f4918b5c9fbc60bae2e5989ac959f73/tests/integrationTests/lib.py#L92-L152
We could try to parallelize this process using modules like [multiprocessing](https://docs.python.org/3/library/multiprocessing.html). Essentially changing the big for-loop to a multiprocess for-loop.

### Selenium e2e tests
We currently use the built-in unittest module as the test suit driver. We could attempt to adopt other testing framework extensions that implement case-level parallelization. A possible candidate would be [testtools](https://pypi.org/project/testtools/) package described [on this StackOverflow post](https://stackoverflow.com/a/17059844), which can parallelize test cases that are based on `unittest.TestCase` that we already satisfy.
https://github.com/Submitty/Submitty/blob/dbee91d58f4918b5c9fbc60bae2e5989ac959f73/tests/e2e/base_testcase.py#L28

## Optimize setup scripts:
Currently, the setup process requires ~20min to execute. The majority of time is spent on `Set up sample course`. This is a tricky one to improve. I have not run a good timing experiment to see what's spending the majority amount of time, the related file is [here](https://github.com/Submitty/Submitty/blob/656f588461e028f3f7e596d1f486cf68646d4cc6/.setup/bin/setup_sample_courses.py).

From what I've observed in the code, this script makes many database requests and system commands. A possible candidate would be to use asyncio to speed up the process. The SQL driver used in this script is SQLAlchemy, which has asyncio based [SQL engine support](https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#sqlalchemy.ext.asyncio.create_async_engine). And we might be able to optimize system calls by delegating them to sub-processes, thus making them effectively async, but this might be not worth the time as most of them are trivial calls, a candidate might be here:
https://github.com/Submitty/Submitty/blob/656f588461e028f3f7e596d1f486cf68646d4cc6/.setup/bin/setup_sample_courses.py#L724-L726
This is invoked in this main loop (line 225):
https://github.com/Submitty/Submitty/blob/656f588461e028f3f7e596d1f486cf68646d4cc6/.setup/bin/setup_sample_courses.py#L222-L227

Something else I've observed in the logs is we are spending ~12mins installing assignment configs (see [here](https://github.com/Submitty/Submitty/runs/2541119628?check_suite_focus=true#step:19:399)), though I have not found the call site to the compilation.

## Other ideas
Currently, there are three disincentive types of e2e tests:
1. Selenium e2e tests [~10min]
2. Cypress e2e tests [~3min]
3. Integration testing [~30min]
It might be ideal to have these three tests to be separated into three jobs. Thus we get three runners running all the tests concurrently. 

This is reasonably easy to implement. Create a setup job that fills the Postgres database with sample data (spending about 20min). Store the Postgres data directory as an artifact/cache, launch the above tests as multiple distinct jobs that relies on the setup job.

This is trivial currently as this only saves ~10mins, but it would come handly as we scale or once we make the setup process trivial.

**Comment by wusatosi** [comment]
> I would focus first on improving the integration test suite to run each suite concurrently. Improving the speed of setup_sample_course is possible, but I would probably want to see this script get cleaned-up as it's quite a chore to edit right now, as well as would be useful to insert timing information first to determine where the slowdowns are, and that certain things could be sped up by making better use of bulk queries vs asyncio.

I agree. Speed-up integration testing would be the easiest with significant speedup.

**Comment by wusatosi** [comment]
I wonder if we can separate integration tests out directly? I don't think integration test need 
1. Any connection with the database 
2. Switching around with the users
It is, as far as I can tell, path sensitive and need some build tools.

**Comment by wusatosi** [comment]
I've been making a paralleled version of the integration test, it's getting about 2x performance boost, possibly because the given system is [dual-core](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources), however some tests are timing out because its performance bounded (see [here](https://github.com/wusatosi/Submitty/runs/2581298659?check_suite_focus=true#step:21:365)).
If we really want to go for speed, we could use the macOS instance which gives us 3-core.

**Comment by wusatosi** [comment]
I have the prototype for the optimization for _integration testing_ and _Selenium e2e testing_, the average time save estimates to be 2x, `~30min -> ~15min` and `~10min -> ~5min`. Which would reduce the time for the e2e CI job from ~1hr to ~40min without separating the _integration testing_ out, for which I would expect an overall time drop from `~40min -> ~30min` with another job running in parallel lasting ~20min.

**Comment by MasterOdin** [comment]
At the moment, I am going to reject trying to use macOS based runner, if just that we only have bandwidth to manage dependency management on one OS (Ubuntu) at the moment, and I don't particularly want to stretch ourselves on this at the moment.

**Comment by MasterOdin** [comment]
One gain for the integration test suite would be to move the compilation step of configure.out from a per-test instance to once per suite run. As part of the "prebuild" step, [build_main_configure.sh](https://github.com/Submitty/Submitty/blob/master/tests/integrationTests/build_main_configure.sh) is run where the compilation step of `configure.out` in it takes a number of seconds. This file is the same for all modules, and so could be safely hoisted to be compiled at the start of run.py to some temp location and then copied into place as needed for each test suite. This would save something like ~16s per module.

**Comment by wusatosi** [comment]
> One gain for the integration test suite would be to move the compilation step of configure.out from a per-test instance to once per suite run. As part of the "prebuild" step, [build_main_configure.sh](https://github.com/Submitty/Submitty/blob/master/tests/integrationTests/build_main_configure.sh) is run where the compilation step of `configure.out` in it takes a number of seconds. This file is the same for all modules, and so could be safely hoisted to be compiled at the start of run.py to some temp location and then copied into place as needed for each test suite. This would save something like ~16s per module.

Thanks for the notice! I will definitely look into this.

**Comment by wusatosi** [comment]
Something else that came up is to use [custom containers](https://github.blog/changelog/2020-09-24-github-actions-private-registry-support-for-job-and-service-containers/) for caching apt packages and external packages like cypress, apache, Nginx, this would not necessarily get a significant speed-up (~2min) but would make our test time more stable as it reduces dependency on external traffics. The first step would move the scripts under `.setup/testing/*` into this private container. 

Though we would need to maintain a docker image on GitHub registry / docker-hub (but this could be automatable).
